File: juno_rs/indicators/kama.rs
use super::MA;
use std::{cmp::min, collections::VecDeque};

pub struct Kama {
    pub value: f64,

    short_alpha: f64,
    long_alpha: f64,

    prices: VecDeque<f64>,
    diffs: VecDeque<f64>,

    t: u32,
    t1: u32,
    t2: u32,
}

impl Kama {
    pub fn new(period: u32) -> Self {
        Self {
            value: 0.0,

            short_alpha: 2.0 / (2.0 + 1.0),
            long_alpha: 2.0 / (30.0 + 1.0),

            prices: VecDeque::with_capacity(period as usize),
            diffs: VecDeque::with_capacity(period as usize),

            t: 0,
            t1: period - 1,
            t2: period,
        }
    }

    pub fn maturity(&self) -> u32 {
        self.t1
    }

    pub fn update(&mut self, price: f64) {
        if self.t > 0 {
            if self.diffs.len() == self.t2 as usize {
                self.diffs.pop_front();
            }
            self.diffs
                .push_back(f64::abs(price - self.prices[self.prices.len() - 1]));
        }

        if self.t == self.t1 {
            self.value = price;
        } else if self.t == self.t2 {
            let diff_sum: f64 = self.diffs.iter().sum();
            let er = if diff_sum == 0.0 {
                1.0
            } else {
                f64::abs(price - self.prices[0]) / diff_sum
            };
            let sc = f64::powf(
                er * (self.short_alpha - self.long_alpha) + self.long_alpha,
                2.0,
            );

            self.value += sc * (price - self.value);
        }

        if self.prices.len() == self.t2 as usize {
            self.prices.pop_front();
        }
        self.prices.push_back(price);
        self.t = min(self.t + 1, self.t2)
    }
}

impl MA for Kama {
    fn update(&mut self, price: f64) {
        self.update(price)
    }

    fn value(&self) -> f64 {
        self.value
    }

    fn maturity(&self) -> u32 {
        self.maturity()
    }
}

#[cfg(test)]
mod tests {
    use super::{super::MA, Kama};

    #[test]
    fn test_kama() {
        let inputs = vec![
            50.25, 50.55, 52.5, 54.5, 54.1, 54.12, 55.5, 50.2, 50.45, 50.24, 50.24, 55.12, 56.54,
            56.12, 56.1, 54.12, 59.54, 54.52,
        ];
        let expected_outputs = vec![
            54.5000, 54.3732, 54.2948, 54.6461, 53.8270, 53.3374, 52.8621, 51.8722, 53.1180,
            54.4669, 55.0451, 55.4099, 55.3468, 55.7115, 55.6875,
        ];
        let mut indicator = Kama::new(4);
        for i in 0..inputs.len() {
            indicator.update(inputs[i]);
            if i >= 3 {
                let diff = f64::abs(indicator.value() - expected_outputs[i - 3]);
                assert!(diff < 0.001, format!("diff is {}", diff));
            }
        }
    }
}

File: juno/indicators/kama.py
from __future__ import annotations

from collections import deque
from decimal import Decimal
from typing import Deque


# Kaufman's Adaptive Moving Average
class Kama:
    value: Decimal = Decimal('0.0')

    _short_alpha: Decimal
    _long_alpha: Decimal

    # _prev_price: Decimal
    _prices: Deque[Decimal]
    _diffs: Deque[Decimal]

    _t: int = -1
    _t1: int
    _t2: int

    def __init__(self, period: int) -> None:
        if period < 1:
            raise ValueError(f'Invalid period ({period})')

        self._short_alpha = Decimal('2.0') / (Decimal('2.0') + 1)
        self._long_alpha = Decimal('2.0') / (Decimal('30.0') + 1)

        self._prices = deque(maxlen=period)
        self._diffs = deque(maxlen=period)

        self._t1 = period - 1
        self._t2 = period

    @property
    def maturity(self) -> int:
        return self._t1

    @property
    def mature(self) -> bool:
        return self._t >= self._t1

    def update(self, price: Decimal) -> Decimal:
        self._t = min(self._t + 1, self._t2)

        if self._t > 0:
            self._diffs.append(abs(price - self._prices[-1]))

        if self._t == self._t1:
            self.value = price
        elif self._t == self._t2:
            diff_sum = sum(self._diffs)
            if diff_sum == 0:
                er = Decimal('1.0')
            else:
                er = abs(price - self._prices[0]) / diff_sum
            sc = (er * (self._short_alpha - self._long_alpha) + self._long_alpha)**2

            self.value += sc * (price - self.value)

        self._prices.append(price)
        return self.value
